### PR TITLE
Offer List #6070

### PR DESCRIPTION
**Pull request by iPR0D1G1**
Offer List #6070

### CODE REVIEW

File: libs/admin/src/lib/admin-panel/admin-panel.module.ts [hunk 1]
```diff
@@ -57,6 +59,7 @@ export const panelRoutes: Routes = [
   { path: 'organization/:orgId', component: OrganizationComponent },
   { path: 'movie/:movieId', component: MovieComponent },
   { path: 'users', component: UsersComponent },
+  { path: 'offers', component: OffersComponent },
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/admin-panel.module.ts:62]: Best would be to create a module for this component and lazyload it. I know we don't do that for the other component of the module, but we plan to improve that progressively : 
```typescript
{ path: 'offers', loadChildren: () => import('./pages/offer/list/list.module').then(m => OfferListModule) }
```

File: libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Component({
+  selector: 'offers',
+  templateUrl: './offers.component.html',
+  styleUrls: ['./offers.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class OffersComponent implements OnDestroy {
+
+  subs: Subscription[] = [];
+  offers = queryChanges.call(this.offerService, {
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts:19]: We usually put a $ at the end of the properties when it's an observable: `offers$`

File: libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    path: 'offers',
+    //@ts-ignore
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts:21]: Create a type for that in the same file and cast the query : 
```typescript
type OfferWithContracts = Offer & { contracts: Contract[] };
const queryOffer: Query<OfferWithContracts> = {
  path: 'offers',
  contracts: offer => ({
    path: 'contracts',
    queryFn: ref => ref.where('offerId', '==', offer.id)
  })
}
@Component({ .... })
export class OfferListComponent {
  offers$ = queryChanges.call(this.offerService, queryOffer);
}
```

File: libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+@Component({
+  selector: 'offers',
+  templateUrl: './offers.component.html',
+  styleUrls: ['./offers.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class OffersComponent implements OnDestroy {
+
+  subs: Subscription[] = [];
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/pages/offers/offers.component.ts:18]: You don't have any subscriptions. Remove it, and the ngOnDestroy

File: libs/admin/src/lib/admin-panel/pages/offers/offers.module.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+@NgModule({
+  declarations: [
+    OffersComponent,
+  ],
+  imports: [
+    CommonModule,
+  ]
+})
+export class OffersModule { }
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/pages/offers/offers.module.ts:14]: If you import the component directly in the AdminPanelModule you don't need a module for the component.
But following the comment above I would like you to lazy load this component, so you need a module, but you'll need to import every dependancy required in the component (`MatTabsModule`, ...) and a `RouterModule.forChild()`

File: libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html [hunk 6]
```diff
@@ -21,8 +21,8 @@
   </button>
 
 
-  <button mat-button [matMenuTriggerFor]="contractsMenu">
-    <mat-icon svgIcon="document"></mat-icon><span matLine>Contracts</span>
+  <button mat-button routerLink="/c/o/admin/panel/offers">
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html:24]: always use a `a` tag when using `routerLink`. Also you might want to add a routerLinkActive class.

Comment (iPR0D1G1) [reply libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html:24]: Just verified the design and there's no need for a routerLinkActive class. the design doesn't specifies which link is currently active.

File: libs/utils/src/lib/pipes/total.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    data.forEach(s => something += +getDeepValue(s, path))
+    return something;
+  }
+}
+
+@NgModule({
+  declarations: [TotalPipe],
+  exports: [TotalPipe]
+})
+export class TotalPipeModule { }
```
Comment (pldespaigne) [inline libs/utils/src/lib/pipes/total.pipe.ts:21]: I also created a similar pipe in `/lib/admin-panel/pages/offers/shell/shell.component.ts` (at the bottom of the file)
Could please delete my pipe, and use yours in the `shell.component.html`

**Review by pldespaigne — CHANGES_REQUESTED** [review]

File: libs/admin/src/lib/admin-panel/pages/offers/list/offer-list.component.scss [hunk 8]
```diff
@@ -0,0 +1,27 @@
+nav[mat-tab-nav-bar] button {
+  border: none;
+  background-color: transparent;
+}
```
[the remark this answers is not in the document]
Comment (iPR0D1G1) [reply libs/admin/src/lib/admin-panel/pages/offers/list/offer-list.component.scss:4]: @GrandSchtroumpf  Told me to do this.

File: libs/utils/src/lib/pipes/join.pipe.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  transform(data: unknown[], path: string = '', seperator: string = ', ') {
+    return data.map(s => typeof s === 'string' ? s : getDeepValue(s, path)).join(seperator)
+  }
+}
+
+@NgModule({
+  declarations: [JoinPipe],
+  exports: [JoinPipe]
+})
+export class JoinPipeModule { }
```
[the remark this answers is not in the document]
Comment (iPR0D1G1) [reply libs/utils/src/lib/pipes/join.pipe.ts:15]: this is intended to be used by @GrandSchtroumpf in the future.

File: libs/admin/src/lib/admin-panel/pages/offers/list/offer-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  ) { }
+
+  /** Dynamic filter of offers for each tab. */
+  applyFilter(filter?: AllOfferStatus) {
+    this.filter.setValue(filter);
+  }
+
+  /* index paramter is unused because it is a default paramter from the filter javascript function */
+  filterByStatus(offer: Offer, index: number, value: AllOfferStatus): boolean {
+    if (value === 'pending') {
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin-panel/pages/offers/list/offer-list.component.ts:71]: Prefer switch case here : 
```typescript
switch (value) {
  case 'pending': return offer.status === value;
  case ...
  default: true;
}
```

File: libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html [hunk 11]
```diff
@@ -21,9 +21,9 @@
   </button>
 
 
-  <button mat-button [matMenuTriggerFor]="contractsMenu">
-    <mat-icon svgIcon="document"></mat-icon><span matLine>Contracts</span>
-  </button>
+  <a mat-button routerLink="/c/o/admin/panel/offers">
```
Comment (GrandSchtroumpf) [inline libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html:24]: I think you only need `routerLink="offers"`

File: libs/contract/src/lib/offer/list/list.component.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  Component, ChangeDetectionStrategy,
+} from '@angular/core';
+
+@Component({
+  selector: 'offers-list',
+  templateUrl: './list.component.html',
+  styleUrls: ['./list.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class ListComponent {
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/offer/list/list.component.ts:11]: Looks like you're not using this component. Can we remove it?

File: libs/utils/src/lib/pipes/total.pipe.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import { NgModule, Pipe, PipeTransform } from '@angular/core';
+import { getDeepValue } from './deep-key.pipe';
+
+@Pipe({ name: 'total' })
+export class TotalPipe implements PipeTransform {
+  transform(data: unknown[], path: string,): number {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/pipes/total.pipe.ts:6]: remove trailing "," in the params

**Review by GrandSchtroumpf — APPROVED** [review]

File: libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html (hunk 11 above)
Comment (iPR0D1G1) [reply libs/admin/src/lib/admin/toolbar-top/toolbar-top.component.html:24]: Bildschirmfoto 2021-06-25 um 11 12 29 AM
We actually need the entire correct url for proper routing.

**Review by pldespaigne — APPROVED** [review]

**Pull request merged**